### PR TITLE
[flask] add /covidcast/csv endpoint

### PR DESCRIPTION
**Pull request by sgratzl**
similar to covidcast-csv-server https://github.com/cmu-delphi/covidcast-csv-server.

as mentioned in the docs with the new streaming and custom endpoints we could add the csv server functionality into the flask api directly.

### CODE REVIEW

File: src/server/endpoints/covidcast.py [hunk 1]
```diff
@@ -170,3 +173,78 @@ def gen(rows):
...
     return p(gen(r))
+
+
+@bp.route("/csv", methods=("GET", "POST"))
+def handle_export():
+    source, signal = request.args.get("signal", "jhu-csse:confirmed_incidence_num").split(":")
+    start_day = request.args.get("start_day", "2020-04-01")
+    end_day = request.args.get("end_day", "2020-09-01")
+    geo_type = request.args.get("geo_type", "county")
+    geo_values = request.args.get("geo_values", "*")
```
Comment (krivard) [inline src/server/endpoints/covidcast.py:184]: Is there reasoning behind these defaults?

File: src/server/endpoints/covidcast.py [hunk 2]
```diff
@@ -170,3 +173,78 @@ def gen(rows):
...
+    # execute query
+    try:
+        r = run_query(p, (str(q), q.params))
+    except Exception as e:
+        raise DatabaseErrorException(str(e))
+
+    # special case for no data to be compatible with the CSV server
+    first_row = next(r, None)
+    if not first_row:
+        return "No matching data found for signal {source}:{signal} " "at {geo} level from {start_day} to {end_day}, as of {as_of}.".format(
```
Comment (krivard) [inline src/server/endpoints/covidcast.py:245]: What's with the `" "`?

File: src/server/endpoints/covidcast.py [hunk 3]
```diff
@@ -170,3 +173,78 @@ def gen(rows):
...
+    q.params["start_day"] = date_to_time_value(start_day)
+    q.params["end_day"] = date_to_time_value(end_day)
+    q.where_geo_pairs("geo_type", "geo_value", [GeoPair(geo_type, True if geo_values == "*" else geo_values)])
+
+    _handle_lag_issues_as_of(q, None, None, date_to_time_value(as_of) if as_of is not None else None)
+
+    # tag as_of in filename, if it was specified
+    as_of_str = "-asof-{as_of}".format(as_of=as_of.isoformat()) if as_of is not None else ""
+    filename = "covidcast-{source}-{signal}-{start_day}-to-{end_day}{as_of}".format(source=source, signal=signal, start_day=start_day.isoformat(), end_day=end_day.isoformat(), as_of=as_of_str)
+    p = CSVPrinter(filename)
```
Comment (krivard) [inline src/server/endpoints/covidcast.py:213]: This is a good idea but we should make sure the covidcast library isn't expecting a universal/static filename before merging

**Review by krivard — COMMENTED** [review]
Small questions

**Comment by sgratzl** [comment]
all of those questions are regarding code that is directly taken from 

https://github.com/cmu-delphi/covidcast-csv-server/blob/main/src/app.py#L44-L89

I just copied them to mimic its behavior

File: src/server/endpoints/covidcast.py (hunk 3 above)
Comment (sgratzl) [reply src/server/endpoints/covidcast.py:213]: I don't understand, this new endpoint is solely for  https://delphi.cmu.edu/covidcast/export/ to replace the covidcast-csv-server

Comment (krivard) [reply src/server/endpoints/covidcast.py:213]: ah, I misunderstood. This is fine then

**Review by krivard — APPROVED** [review]
👍  as a separate covidcast/export endpoint

**Pull request merged**